File: java/07 - Demo 3 - Refactoring/src/test/java/introduceparameterobject/FlowBetweenTest.java
package introduceparameterobject;

import static org.junit.Assert.assertEquals;

import java.util.Calendar;
import java.util.Date;

import org.junit.Before;
import org.junit.Test;

public class FlowBetweenTest {

    private Account account;

    @Before
    public void setUp() throws Exception {
        account = new Account();
    }

    @Test
    public void beforeIntervalIsExcluded() {
        Entry beforeInterval = new Entry(10.0, createDate(1, 3));
        account.getEntries().add(beforeInterval);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(0.0, result, 0.0);
    }

    @Test
    public void onStartOfIntervalIsIncluded() {
        Entry startOfInterval = new Entry(10.0, createDate(1, 4));
        account.getEntries().add(startOfInterval);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(10.0, result, 0.0);
    }

    @Test
    public void middleOfIntervalIsIncluded() {
        Entry middleOfInterval = new Entry(10.0, createDate(2, 2));
        account.getEntries().add(middleOfInterval);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(10.0, result, 0.0);
    }

    @Test
    public void onEndOfIntervalIsIncluded() {
        Entry endOfInterval = new Entry(10.0, createDate(2, 4));
        account.getEntries().add(endOfInterval);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(10.0, result, 0.0);
    }

    @Test
    public void afterIntervalIsExcluded() {
        Entry before = new Entry(10.0, createDate(2, 5));
        account.getEntries().add(before);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(0.0, result, 0.0);
    }

    @Test
    public void multipleEntriesInIntervalAreAccumulated() {
        Entry middleOfInterval = new Entry(10.0, createDate(2, 2));
        Entry endOfInterval = new Entry(10.0, createDate(2, 4));
        account.getEntries().add(middleOfInterval);
        account.getEntries().add(endOfInterval);

        double result = account.getFlowBetween(createDate(1, 4), createDate(2, 4));
        assertEquals(20.0, result, 0.0);
    }

    private Date createDate(int month, int day) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(2007, month, day);
        return calendar.getTime();
    }


}

File: csharp/07 - Demo 3 - Refactoring/Refactoring.Tests/IntroduceParameterObject/FlowBetweenTest.cs
using System;
using Xunit;
using FluentAssertions;
using Refactoring.IntroduceParameterObject;

namespace Refactoring.Tests.IntroduceParameterObject
{
    public class FlowBetweenTest
    {
        private Account _account;

        public FlowBetweenTest()
        {
            _account = new Account();
        }

        [Fact]
        public void BeforeIntervalIsExcluded()
        {
            var beforeInterval = new Entry(10.0, new DateTime(2007, 1, 3));
            _account.Entries.Add(beforeInterval);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(0.0);
        }

        [Fact]
        public void OnStartOfIntervalIsIncluded()
        {
            var startOfInterval = new Entry(10.0, new DateTime(2007, 1, 4));
            _account.Entries.Add(startOfInterval);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(10.0);
        }

        [Fact]
        public void MiddleOfIntervalIsIncluded()
        {
            var middleOfInterval = new Entry(10.0, new DateTime(2007, 2, 2));
            _account.Entries.Add(middleOfInterval);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(10.0);
        }

        [Fact]
        public void OnEndOfIntervalIsIncluded()
        {
            var endOfInterval = new Entry(10.0, new DateTime(2007, 2, 4));
            _account.Entries.Add(endOfInterval);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(10.0);
        }

        [Fact]
        public void AfterIntervalIsExcluded()
        {
            var before = new Entry(10.0, new DateTime(2007, 2, 5));
            _account.Entries.Add(before);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(0.0);
        }

        [Fact]
        public void MultipleEntriesInIntervalAreAccumulated()
        {
            var middleOfInterval = new Entry(10.0, new DateTime(2007, 2, 2));
            var endOfInterval = new Entry(10.0, new DateTime(2007, 2, 4));
            _account.Entries.Add(middleOfInterval);
            _account.Entries.Add(endOfInterval);

            var result = _account.GetFlowBetween(new DateTime(2007, 1, 4), new DateTime(2007, 2, 4));
            result.Should().Be(20.0);
        }
    }
}